// twitch-pubsub-client/lib/index.js
var __tsu = {
    cache: [],
    defineExport: function (name, def) {
        Object.defineProperty(exports, name, def);
        this.cache.push({ name: name, def: def });
    },
    redefineExports: function () {
        this.cache.forEach(function (exp) {
            Object.defineProperty(exports, exp.name, exp.def);
        });
    }
};
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LogLevel = exports.PubSubWhisperMessage = exports.PubSubSubscriptionMessage = exports.PubSubRedemptionMessage = exports.PubSubChatModActionMessage = exports.PubSubBitsBadgeUnlockMessage = exports.PubSubBitsMessage = exports.SingleUserPubSubClient = exports.BasicPubSubClient = exports.PubSubListener = exports.PubSubClient = void 0;
var tslib_1 = require("tslib");
/* eslint-disable filenames/match-exported */
var shared_utils_1 = require("@d-fischer/shared-utils");
var PubSubClient_1 = require("./PubSubClient");
__tsu.defineExport("PubSubClient", { enumerable: true, get: function () { return PubSubClient_1.PubSubClient; } });
/** @deprecated Use the named export `PubSubClient` instead. */
var DeprecatedPubSubClient = shared_utils_1.deprecateClass(PubSubClient_1.PubSubClient, "[twitch-pubsub-client] The default export has been deprecated. Use the named export instead:\n\n\timport { PubSubClient } from 'twitch-pubsub-client';");
/** @deprecated Use the named export `PubSubClient` instead. */
exports.default = DeprecatedPubSubClient;
var PubSubListener_1 = require("./PubSubListener");
__tsu.defineExport("PubSubListener", { enumerable: true, get: function () { return PubSubListener_1.PubSubListener; } });
var BasicPubSubClient_1 = require("./BasicPubSubClient");
__tsu.defineExport("BasicPubSubClient", { enumerable: true, get: function () { return BasicPubSubClient_1.BasicPubSubClient; } });
var SingleUserPubSubClient_1 = require("./SingleUserPubSubClient");
__tsu.defineExport("SingleUserPubSubClient", { enumerable: true, get: function () { return SingleUserPubSubClient_1.SingleUserPubSubClient; } });
tslib_1.__exportStar(require("./PubSubPacket"), exports);
var PubSubBitsMessage_1 = require("./Messages/PubSubBitsMessage");
__tsu.defineExport("PubSubBitsMessage", { enumerable: true, get: function () { return PubSubBitsMessage_1.PubSubBitsMessage; } });
var PubSubBitsBadgeUnlockMessage_1 = require("./Messages/PubSubBitsBadgeUnlockMessage");
__tsu.defineExport("PubSubBitsBadgeUnlockMessage", { enumerable: true, get: function () { return PubSubBitsBadgeUnlockMessage_1.PubSubBitsBadgeUnlockMessage; } });
var PubSubChatModActionMessage_1 = require("./Messages/PubSubChatModActionMessage");
__tsu.defineExport("PubSubChatModActionMessage", { enumerable: true, get: function () { return PubSubChatModActionMessage_1.PubSubChatModActionMessage; } });
var PubSubRedemptionMessage_1 = require("./Messages/PubSubRedemptionMessage");
__tsu.defineExport("PubSubRedemptionMessage", { enumerable: true, get: function () { return PubSubRedemptionMessage_1.PubSubRedemptionMessage; } });
var PubSubSubscriptionMessage_1 = require("./Messages/PubSubSubscriptionMessage");
__tsu.defineExport("PubSubSubscriptionMessage", { enumerable: true, get: function () { return PubSubSubscriptionMessage_1.PubSubSubscriptionMessage; } });
var PubSubWhisperMessage_1 = require("./Messages/PubSubWhisperMessage");
__tsu.defineExport("PubSubWhisperMessage", { enumerable: true, get: function () { return PubSubWhisperMessage_1.PubSubWhisperMessage; } });
var logger_1 = require("@d-fischer/logger");
__tsu.defineExport("LogLevel", { enumerable: true, get: function () { return logger_1.LogLevel; } });
module.exports = exports = DeprecatedPubSubClient;
Object.defineProperty(exports, "__esModule", { value: true });
__tsu.redefineExports();
/** @deprecated Use the named export `PubSubClient` instead. */
exports.default = DeprecatedPubSubClient;
