// twitch-webhooks/lib/index.js
var __tsu = {
    cache: [],
    defineExport: function (name, def) {
        Object.defineProperty(exports, name, def);
        this.cache.push({ name: name, def: def });
    },
    redefineExports: function () {
        this.cache.forEach(function (exp) {
            Object.defineProperty(exports, exp.name, exp.def);
        });
    }
};
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Subscription = exports.SimpleAdapter = exports.ReverseProxyAdapter = exports.LegacyAdapter = exports.EnvPortAdapter = exports.ConnectionAdapter = exports.WebHookListener = void 0;
/* eslint-disable filenames/match-exported */
var shared_utils_1 = require("@d-fischer/shared-utils");
var WebHookListener_1 = require("./WebHookListener");
__tsu.defineExport("WebHookListener", { enumerable: true, get: function () { return WebHookListener_1.WebHookListener; } });
/** @deprecated Use the named export `WebHookListener` instead. */
var DeprecatedWebHookListener = shared_utils_1.deprecateClass(WebHookListener_1.WebHookListener, "[twitch-webhooks] The default export has been deprecated. Use the named export instead:\n\n\timport { WebHookListener } from 'twitch-webhooks';");
/** @deprecated Use the named export `WebHookListener` instead. */
exports.default = DeprecatedWebHookListener;
var ConnectionAdapter_1 = require("./Adapters/ConnectionAdapter");
__tsu.defineExport("ConnectionAdapter", { enumerable: true, get: function () { return ConnectionAdapter_1.ConnectionAdapter; } });
var EnvPortAdapter_1 = require("./Adapters/EnvPortAdapter");
__tsu.defineExport("EnvPortAdapter", { enumerable: true, get: function () { return EnvPortAdapter_1.EnvPortAdapter; } });
var LegacyAdapter_1 = require("./Adapters/LegacyAdapter");
__tsu.defineExport("LegacyAdapter", { enumerable: true, get: function () { return LegacyAdapter_1.LegacyAdapter; } });
var ReverseProxyAdapter_1 = require("./Adapters/ReverseProxyAdapter");
__tsu.defineExport("ReverseProxyAdapter", { enumerable: true, get: function () { return ReverseProxyAdapter_1.ReverseProxyAdapter; } });
var SimpleAdapter_1 = require("./Adapters/SimpleAdapter");
__tsu.defineExport("SimpleAdapter", { enumerable: true, get: function () { return SimpleAdapter_1.SimpleAdapter; } });
var Subscription_1 = require("./Subscriptions/Subscription");
__tsu.defineExport("Subscription", { enumerable: true, get: function () { return Subscription_1.Subscription; } });
module.exports = exports = DeprecatedWebHookListener;
Object.defineProperty(exports, "__esModule", { value: true });
__tsu.redefineExports();
/** @deprecated Use the named export `WebHookListener` instead. */
exports.default = DeprecatedWebHookListener;
